libs/filesystem/src/main/java/org/padaiyal/utilities/filesystem/FileSystemUtility.java

```diff
@@ -0,0 +1,692 @@
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.function.BiConsumer;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.padaiyal.utilities.I18nUtility;
+import org.padaiyal.utilities.PropertyUtility;
+import org.padaiyal.utilities.filesystem.abstractions.WatchServiceRegisteringFileVisitor;
+import org.padaiyal.utilities.filesystem.abstractions.WatchServiceRegistrationInfo;
+
+/**
+ * File utility library for creating, modifying, deleting and monitoring files and directories.
+ */
+public class FileSystemUtility {
+
+  /**
+   * Logger object to log information and errors.
+   */
+  private static final Logger logger = LogManager.getLogger(FileSystemUtility.class);
+
+  /**
+   * Map of paths to the WatchRegistrationInfo, used to store events to track for each path.
+   */
+  private static final ConcurrentHashMap<String, WatchServiceRegistrationInfo>
+      pathToWatchServiceRegistrationInfoMap = new ConcurrentHashMap<>();
+
+  /**
+   * Map that stores the path that triggered an event and the event type they generated.
+   */
+  private static final ConcurrentHashMap<String, Queue<Kind<?>>> pathToTriggeredEventMap =
+      new ConcurrentHashMap<>();
+
+  /**
+   * Thread object that visits all files to look for events.
+   */
+  private static Thread backgroundWatchServiceRegisteringThread;
+
+  /**
+   * Runnable used by the backgroundWatchServiceRegisteringThread.
+   */
+  public static final Runnable backgroundWatchServiceRegisteringThreadRunnable = () -> {
+
+    while (Boolean.parseBoolean(
+            PropertyUtility.getProperty(
+                "FileSystemUtility.backgroundThread.watchService.switch"))) {
+      HashMap<String, WatchServiceRegistrationInfo>
+          pathToWatchServiceRegistrationInfoMapCopy =
+          new HashMap<>(pathToWatchServiceRegistrationInfoMap);
+
+      // Iterate through all registered watch services and check if any events are
+      // generated
+      pathToWatchServiceRegistrationInfoMapCopy.keySet()
+          .stream()
+          .sorted()
+          .forEach(
+              (subPath) -> {
+
+                WatchServiceRegistrationInfo subPathWatchServiceInfo =
+                        pathToWatchServiceRegistrationInfoMapCopy
+                                .get(subPath);
+
+                if (subPathWatchServiceInfo != null) {
```

Comment:
I'm presuming you added this if clause to fix this [issue](https://github.com/padaiyal/jMonocle/pull/11#discussion_r1118310444) ?
We'd need to either log a warning/error or raise an exception when `subPathWatchServiceInfo` is null.

Reply:
which one would be preferable ? do we prefer that it fail or is it more of a warning ?

Reply:
I would vote to log a warning message with subPath and pathToWatchServiceRegistrationInfoMapCopy.
And skip the current iteration and continue.

Reply:
sounds good. Will add a warning.